app/Actions/Settings/Login.php

```diff
@@ -58,29 +92,36 @@ public function do(Request $request)
 			$user = User::query()->findOrFail($id);
 
 			if ($user->is_locked) {
-				Logs::notice(__METHOD__, __LINE__, 'Locked user (' . $user->username . ') tried to change their identity from ' . $request->ip());
-				throw new JsonError('Locked account!');
+				Logs::notice(__METHOD__, __LINE__, 'Locked user (' . $user->username . ') tried to change their identity from ' . $ip);
+				throw new UnauthorizedException('Account is locked');
 			}
 
-			if (User::query()->where('username', '=', $request['username'])->where('id', '!=', $id)->count()) {
-				Logs::notice(__METHOD__, __LINE__, 'User (' . $user->username . ') tried to change their identity to ' . $request['username'] . ' from ' . $request->ip());
-
-				throw new JsonError('Username already exists.');
+			if (User::query()->where('username', '=', $username)->where('id', '!=', $id)->count()) {
+				Logs::notice(__METHOD__, __LINE__, 'User (' . $user->username . ') tried to change their identity to ' . $username . ' from ' . $ip);
+				throw new ConflictingPropertyException('Username already exists.');
 			}
 
-			if ($user->username == $oldUsername && Hash::check($oldPassword, $user->password)) {
-				Logs::notice(__METHOD__, __LINE__, 'User (' . $user->username . ') changed their identity for (' . $request['username'] . ') from ' . $request->ip());
-
-				$user->username = $request['username'];
-				$user->password = bcrypt($request['password']);
-
-				return $user->save();
+			// TODO: This looks suspicious.
+			// Users can only change the username/password of their own
+			// account and must be authenticated in order to do so.
+			// (See above, we use `AccessControl::id()` and query for the
+			// currently authenticated user).
+			// The user name of the currently authenticated user is visible on
+			// the GUI anyway.
+			// Why do we re-check whether users have been able to correctly
```

Comment:
Most likely legacy stuff from when there was no multi user support.

Reply:
Yes, that is what I assumed, too. But what exactly do you want to tell me by your comment? What is your intention?

IMHO, there are three (may be four) areas in the code which are dangerous and need improvement:

 1. User and Session Management: See also [my comment in this tread](https://github.com/LycheeOrg/Lychee/pull/1217#discussion_r825435745).
 2. Web updater/installer: If it works, it works and everything is fine. If it fails (in particular, if dependencies via composer must be installed), then the user has no idea what went wrong. IMHO, at the moment this component does more harm than it helps. Moreover, the code is really wicked and incomprehensible, also see [this comment](https://github.com/LycheeOrg/Lychee/pull/1217#discussion_r825437068).
 3. Exif data and file types: (You might count this as two, if you wish). The Exiftool returns a nice and type-safe object with extracted exif data. However, our code copies all the values into an unsafe, associative array and passes around that array. Also the array is changed in some non-obvious ways while it is passed around.  Moreover, we have a lot (i mean A LOT) of duplicated code which decides if we support a certain type of file or not (and of course the decision logic is always slightly different)

However, I did not come up with this code. When I added proper exception handling to our code I tried to work around these areas as well as I could without touching too much in these areas. IMHO, these three components needs to be rewritten from scratch. They are beyond "repairability".

IMHO, this should be three different, future PRs. Luckily, the impact of these PRs while be much more locally concentrated than this PR.

Reply:
1. User and Session management are indeed in need of improvement. I do hope to be able to refactor that a bit at least the front-end side.
2. I am open to having those rewritten from scratch.
I tried my best to provide an installer which works in similar fashion as the one provided with Lychee v3, that was one of the big complaint when moving to v4: there was the need of some commands to be executed with SSH which obviously is not really doable in the case of FTP webhosting solutions without command line access.
3. No complaints there either. I am always open to code improvement.